src/library/scala/Option.scala

```diff
@@ -323,6 +323,43 @@ sealed abstract class Option[+A] extends Product with Serializable {
   final def zip[A1 >: A, B](that: Option[B]): Option[(A1, B)] =
     if (isEmpty || that.isEmpty) None else Some((this.get, that.get))
 
+  /** Converts an Option of a pair into an Option of the first element and an Option of the second element.
+    *
+    *  @tparam A1    the type of the first half of the element pair
+    *  @tparam A2    the type of the second half of the element pair
+    *  @param asPair an implicit conversion which asserts that the element type
+    *                of this Option is a pair.
+    *  @return       a pair of Options, containing, respectively, the first and second half
+    *                of the element pair of this Option.
+    */
+  final def unzip[A1, A2](implicit asPair: A => (A1, A2)): (Option[A1], Option[A2]) = {
+    if (isEmpty)
+      (None, None)
+    else {
+      val e = asPair(this.get)
+      (Some(e._1), Some(e._2))
```

Comment:
A pattern match might be preferable to read?
```scala
  val (e1, e2) = asPair(this.get)
  (Some(e1), Some(e2))
```

Reply:
I used the same style as in [ArrayOps.scala](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/ArrayOps.scala#L1153) 

```
  def unzip[A1, A2](implicit asPair: A => (A1, A2), ct1: ClassTag[A1], ct2: ClassTag[A2]): (Array[A1], Array[A2]) = {
    val a1 = new Array[A1](xs.length)
    val a2 = new Array[A2](xs.length)
    var i = 0
    while (i < xs.length) {
      val e = asPair(xs(i))
      a1(i) = e._1
      a2(i) = e._2
      i += 1
    }
    (a1, a2)
  }
```

Reply:
If we change here, should we change the other classes as well?

Reply:
> I used the same style as in [ArrayOps.scala](https://github.com/scala/scala/blob/2.13.x/src/library/scala/collection/ArrayOps.scala#L1153)

Ok, good to know.

That's in a `while` loop, so maybe it's that way for a performance, but I would hope not.  Regardless, performance isn't necessary for `Option` here as it is for `Array`.

Reply:
Offhand I don't recall what the bytecode for the pattern match looks like, but if it's any less efficient, then I would suggest letting efficiency trump readability.

Reply:
The tree generated with `val (x, y) = (a, b)` is quite silly:

```scala
scala> val (x, y) = (1, 2)
[[syntax trees at end of                    parser]] // <console>
/* ... */
  <synthetic> <artifact> private[this] val x$1 = scala.Tuple2(1, 2): @scala.unchecked match {
    case scala.Tuple2((x @ _), (y @ _)) => scala.Tuple2(x, y)
  };
  val x = x$1._1;
  val y = x$1._2
/* ... */

```

but `-opt:l:method` is smart enough to eliminate that allocation.

Reply:
There was definitely a bug ticket about that extra allocation, and whether or not it should be the compiler or optimiser's job to get rid of it.

For reference, the pattern match has *way* worse bytecode without optimisation: https://gist.github.com/NthPortal/4d2ce4a7882709a5b1bbcba1e677f17e

Reply:
and the relevant issue about the extra allocation in the generated bytecode: scala/bug#10767

Reply:
If performance is the measure for rejecting, than would microbenchmark be the measure?  I won't pretend to be an avid bytecode reader.    The bytecode in the gist is a good distillation of tuples v pattern match, generally, but I'm skeptical that it's a fair representation of the specific code in `unzip` and `unzip3`.  I'd rather benchmark this code.

Reply:
So running some benchmarks in JMH gives some surprising comparisons given the intuitions above about bytecode.  The results show the `None` cases are slower with pattern match by a few percent, but the pattern match is about 10% faster with the pattern match for `Some` values.  I suppose combining the two paths it would be a wash, and depend more on the weighting of the two cases in the wild.

In a way, it's not real surprising since this isn't code with loop complexity, and intellectually it doesn't seem like pattern matches and tuple accessors are all that different.

|           |Tuple access (ops/s)|Pattern match (ops/s)|Change (%)|
|-----------|--------------------|---------------------|----------|
|Some.unzip |        44,421,016.538|         49,801,192.979|     +12.1|
|None.unzip |       119,085,275.042|        114,705,650.786|      -3.7|
|Some.unzip3|        36,159,582.118|         42,635,752.994|     +17.9|
|None.unzip3|       103,887,278.669|        97,433,789.205 |      -6.2|

I ran 4 trials alternating between the control (tuple accessors) and the experiment (pattern match) by running 4 forked benchmarks per trial of 16 iterations with 8 warm-up iterations:

    > jmh:run -f 4 -wi 8 -i 16 scala.OptionUnzipBenchmark

And here's the benchmark code:

```scala
package scala

import org.openjdk.jmh.annotations.Benchmark

class OptionUnzipBenchmark {

  @Benchmark
  def scalaOptionUnzip3None = None.unzip3

  @Benchmark
  def scalaOptionUnzip3Some = Some(("foo", "bar", "buz")).unzip3

  @Benchmark
  def scalaOptionUnzipNone = None.unzip

  @Benchmark
  def scalaOptionUnzipSome = Some(("foo", "bar")).unzip
}
```

Reply:
@ashawley by "pattern match" do you mean the `val (a, b) = asPair(get)` destructuring or a full-fledged pattern match on `this`?

Reply:
Yeah, I'm calling destructuring a tuple _a la_ `val (a, b) = asPair(get)` a pattern match.

Reply:
ah, ok. I'm honestly quite surprised that there's a difference in the `None` case; it _seems_ that it shouldn't matter.

Reply:
thx @ashawley, good enough for me

Reply:
Thanks for the benchmark. IMHO I don't think destructing brings more readability, and the performance difference of more than 10% also needs to be justified. None case is quite strange and should probably worth looking into.

Reply:
I can't explain the None case.  I brought up performance as a guess for the rationale for the code you copied from Array.  Others had suggested that performance was important in this case.  I ran the benchmarks only to show that the performance argument was specious, but I'm not an expert.  The benchmarks are anecdotal, so there's a risk they don't reflect actual performance.  Even the improvement shown in the results, is less significant than the variance of the iterations.

I won't argue why destructuring bind is more readable.  I assumed that your thumbs up of Seth's comment meant your only concern for keeping it was performance.

Reply:
I do think performance is important than readability, especially in core framework collection package as it is critical even it is just 10% difference. Our banks here use Scala for processing market data within ticks and cannot afford delayed data because that would cause financial loss.